apis/replication.storage/v1alpha1/volumereplicationclass_webhook.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2022 The Kubernetes-CSI-Addons Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"errors"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var volumereplicationclasslog = logf.Log.WithName("volumereplicationclass-resource")
+
+func (r *VolumeReplicationClass) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/validate-replication-storage-openshift-io-v1alpha1-volumereplicationclass,mutating=false,failurePolicy=fail,sideEffects=None,groups=replication.storage.openshift.io,resources=volumereplicationclasses,verbs=create;update,versions=v1alpha1,name=vvolumereplicationclass.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &VolumeReplicationClass{}
+
+// ValidateCreate implements webhook.Validator. Validates the creation of
+// VolumeReplicationClass and does not allow empty provisioner name.
+func (r *VolumeReplicationClass) ValidateCreate() error {
+	volumereplicationclasslog.Info("validate create", "name", r.Name)
+
+	if r.Spec.Provisioner == "" {
+		return errors.New("provisioner name not provided")
+	}
+
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator. Validates the updated
+// VolumeReplicationClass and does not allow changes to existing spec object.
```

Comment:
If it does not allow any changes to the `.Spec` part, why not use `reflect.DeepEqual` for that? No need to compare `.Provisioner` and `.Parameters` separately.

Reply:
not comparing spec because for now we have only two fields in spec but later we might add new fields which are suppose to be editable also that the reason to compare it explicitly